apps/emqx/src/emqx_mqtt_caps.erl

```diff
@@ -105,17 +105,17 @@ do_check_pub(_Flags, _Caps) ->
     emqx_types:topic() | emqx_types:share(),
     emqx_types:subopts()
 ) ->
-    ok_or_error(emqx_types:reason_code()).
+    ok
+    | {ok, emqx_types:reason_code()}
+    | {error, emqx_types:reason_code()}.
 check_sub(ClientInfo = #{zone := Zone}, Topic, SubOpts) ->
     Caps = emqx_config:get_zone_conf(Zone, [mqtt]),
     Flags = #{
-        %% TODO: qos check
-        %% (max_qos_allowed, Map) ->
-        %% max_qos_allowed => maps:get(max_qos_allowed, Caps, 2),
         topic_levels => emqx_topic:levels(Topic),
         is_wildcard => emqx_topic:wildcard(Topic),
         is_shared => erlang:is_record(Topic, share),
-        is_exclusive => maps:get(is_exclusive, SubOpts, false)
+        is_exclusive => maps:get(is_exclusive, SubOpts, false),
+        qos => maps:get(qos, SubOpts, 0)
```

Comment:
nit, default `0` here is unnecessary, it is mandatory according to the protocol.